---
title: "smFISH analysis"
author: "Lorena Garcia-Perez (lorena.garcia-perez@crick.ac.uk)"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"

output:
  pdf_document:
    latex_engine: xelatex

knit: (function(inputFile, encoding) { 
      out_dir <- "docs";
      rmarkdown::render(inputFile,
                        encoding = encoding, 
                        output_file = file.path(dirname(inputFile),
                                      out_dir, "smFISH_analysis.pdf")) })
---

Main R script to determine:
    RNA copy number per cell
    RNA density number per cell
    Transcriptional properties
Based on the output .csv table obtained with the Python script  
  
Load libraries  
  
```{r, message = FALSE, warning = FALSE}
library(ggplot2)
library (ggbeeswarm)
require(plyr) # Am I using this in the end?
library(tidyverse)
# library(tibble) #for add_column etc
library(knitr)
library(rmarkdown)
library(pander)
```
  
```{r}
opts_chunk$set(panderOptions('table.split.table', 80),
               panderOptions('table.alignment.default', 'left'),
               panderOptions('table.alignment.rownames', 'right'))

```
  
Read data and calculate RNA copy number and concentration (== density)  
  
```{r child = "load_clean.Rmd"}
render("load_clean.Rmd")
```
  
```{r}
copy.density.cell <-
  file %>%
  mutate(cell.area = cell.area * pixel.size) %>%
    # Convert cell area in pixels to cell area in µm^2
  group_by(channel, day, image, cell.id, cell.area) %>%
  summarise(copy = n()) %>% # RNA copy number per cell
  mutate(density = round(copy / cell.area, 3))
    # RNA densities in terms of RNA molecules per µm^2
pander(head(copy.density.cell), style = "rmarkdown")
```
  
To determine which spots can be considered transcription spots,
we need to know the area and maximum intensity values for all the spots
we normalize according to each channel, day and image  
  
```{r}
TX.threholds <-
  file %>%
  group_by(channel, day, image) %>%
    # Choose threholds specifically for each image and channel
  summarise(MaxInt.median = median(max.int), MaxInt.mad = mad(max.int),
            area.median = median(area), area.mad = mad(area)) %>%
  mutate(MaxInt.thres = MaxInt.median + 3 * MaxInt.alpha * MaxInt.mad,
         area.thres = area.median + 3 * area.alpha * area.mad)
pander(head(TX.threholds), style = "rmarkdown")
```
  
```{r}
TX.number.cell <-
  file %>%
  group_by(channel, day, image) %>%
  left_join(TX.threholds) %>%
  mutate(is.TX.spot = max.int > MaxInt.thres & area > area.thres) %>%
  group_by(channel, day, image, cell.id) %>%
  summarise(TX.number = sum(is.TX.spot))
pander(head(TX.number.cell), style = "rmarkdown")

```
  
```{r}
TX.mol <-
  file %>%
  group_by(channel, day, image) %>%
  left_join(TX.threholds) %>%
  mutate(is.TX.spot = max.int > MaxInt.thres & area > area.thres) %>%
  mutate(meanInt.median = median(mean.int)) %>%
  filter(is.TX.spot == TRUE) %>%
  mutate(probe.loc = channel) %>%
  mutate(probe.loc = as.numeric(recode(probe.loc,
                            "0"=correction.factor[1],
                            "1"=correction.factor[2],
                            "2"=correction.factor[3]))) %>%
  mutate(TX.mol = mean.int * area / (probe.loc * meanInt.median * area.median))
TX.mol$channel <- as.factor(TX.mol$channel)
levels(TX.mol$channel) <- genes
pander(head(TX.mol), style = "rmarkdown")
```
  
  
### 2. Main datasets for single cell data:  
  
```{r}
single.cell.data <-
  inner_join(copy.density.cell, TX.number.cell) %>%
  ungroup()
single.cell.data$channel <- as.factor(single.cell.data$channel)
levels(single.cell.data$channel) <- genes
pander(head(single.cell.data), style = "rmarkdown")
```
  
```{r}
single.cell.expressing.data <-
  single.cell.data %>%
  filter(density >= 1)
pander(head(single.cell.expressing.data), style = "rmarkdown")
```
  
```{r}
single.cell.sox2.expressing.data <-
  # This could be combined within single.cell.progenitors.data with filter_if()?
  single.cell.data %>%
  filter(channel == progenitors.gene) %>%
  filter(density >= 1) # Require sox2 density >= 1 molecule / µm^2
pander(head(single.cell.sox2.expressing.data), style = "rmarkdown")
```
  
```{r}
single.cell.progenitors.data <-
  # Select progenitors according to sox2 density >= 1 molecule / µm^2
  semi_join(single.cell.data, single.cell.sox2.expressing.data,
            by = c("day", "image", "cell.id", "cell.area"))
pander(head(single.cell.progenitors.data), style = "rmarkdown")
```
  
```{r}
# This is the more complete, non subseted, dataset
# But, if a cell has two TX spots, it will appear on two rows and so on
withTX.single.cell.data <- 
  TX.mol %>%
  select (channel, day, image, cell.id, spot.id, TX.mol) %>% 
  full_join(single.cell.data) %>%
  arrange(channel, day, image, cell.id) %>%
  replace_na(list(TX.mol = 0))
pander(head(withTX.single.cell.data), style = "rmarkdown")
```
  
```{r}
two.alleles.data <-
  TX.mol %>%
  select(channel, day, image, cell.id, TX.mol) %>%
  group_by(channel, day, image, cell.id) %>%
  mutate(freq = n()) %>% 
  ungroup() %>%
  filter(freq == 2) %>%
  select(-freq) %>%
  mutate(TX.id = rep(c(1,2), length.out = length(.$channel))) %>%
    # Create a column with appropiate number of rows
  spread(TX.id, TX.mol) %>%
  rename(TX.id.1 = "1", TX.id.2 = "2")
two.alleles.data$channel <- as.factor(two.alleles.data$channel )
levels(two.alleles.data$channel) <- genes
pander(head(two.alleles.data), style = "rmarkdown")
```
  
  
### Mean properties  
  
single.cell.progenitors.data is used, but results with single.cell.data are
quite similar  
  
```{r child = "func.Rmd"}
render("func.Rmd")
```
  
```{r}
mean.properties.data <-
  tibble(day = character(), gene = character(), prob.TX = as.numeric(),
         mean.copy = as.numeric(), mean.nPol = as.numeric())

for (d in days) {
  for (g in genes) {
    mean.properties.data <-
      add_row(mean.properties.data,
              day = d, gene = g,
              prob.TX =
                round(prob.TX.fn(single.cell.progenitors.data, d, g, ploidy),3),
              mean.copy =
                round(mean.copy.fn(single.cell.progenitors.data, d, g),3),
              mean.nPol =
                round(mean.nPol.fn(TX.mol, d, g),3))
  }
}
mean.properties.data <- mean.properties.data %>% replace_na(list(mean.nPol = 0))

mean.properties.data <-
  mean.properties.data %>%
  mutate(vPol = vPol.chosen, gene.length = rep(genes.length,3),
         ploidy = ploidy) %>%
  mutate(TX.rate = mean.nPol * vPol / gene.length) %>%
  mutate(deg.rate = prob.TX * TX.rate * ploidy / mean.copy) %>%
  mutate(half.life = log(2)/deg.rate) #in min
pander(head(mean.properties.data), style = "rmarkdown")
```
  
  
## 4.  
  
```{r}
prob.TX.pairs.data <-
  mean.properties.data %>%
  select(day, gene, prob.TX) %>%
  rename(gene.1 = gene, prob.TX.1 = prob.TX) %>%
  mutate(gene.2 = gene.1) %>%
  expand(nesting(day, gene.1, prob.TX.1), gene.2) %>%
  group_by(day) %>%
  mutate(prob.TX.2 = rep(unique(prob.TX.1), times = 3)) %>%
  filter(gene.1 < gene.2) # To get unique combinations
pander(head(prob.TX.pairs.data), style = "rmarkdown")
```
  
## TO BE FINISHED...  

for (d in days) {
  for (g in genes) {
    prob.TX.pairs.data$prob.TX.both <-
      add_row(prob.TX.pairs.data$prob.TX.both,
              prob.TX.pairs.fn())
              
  
```{r}
prob.TX.pairs.fn <-
  function(data, day.chosen, gene.chosen.1, gene.chosen.2,ploidy) {
  # day.chosen and gene.schosen must be input as character (for example, "4"
  # and "Olig2")
  # ploidy must be input as numeric
  day.gene.TX.subset <-
    filter(data, day == day.chosen &
                (channel == gene.chosen.1 | channel == gene.chosen.2))
  dual.TX <-
    day.gene.TX.subset %>%
    select(channel, image, day, cell.id, TX.number) %>%
    filter (TX.number != 0) %>%
    group_by(day, image, cell.id) %>%
    mutate(freq = n()) %>% 
    ungroup() %>%
    filter(freq == 2) %>%
    select(-freq) %>%
    spread(channel, TX.number) %>%
    
  return(nrow(dual.TX)/(nrow(day.gene.TX.subset)))
    # Here I do not make use of the ploidy concept...
}
```
  
two.alleles.data$channel <- as.factor(two.alleles.data$channel )
levels(two.alleles.data$channel) <- genes
  
TX.probability.all[,6] <- TX.probability.all$TX.prob.pair /
  (TX.probability.all$TX.prob.oneC * TX.probability.all$TX.prob.otherC)
colnames(TX.probability.all)[6] <- c("extrinsic.noise")
  
## To measure the influence of extrinsic noise:  
## THIS SHOULD BE DONE WITH A CONSTITUTIVE GENE TO COMPARE!!!  
  
###########################################################################  
  
```{r}
genes.pairwise <-
  tibble(genes) %>%
  rename(gene.1 = genes) %>%
  mutate(gene.2 = gene.1) %>%
  expand(gene.1, gene.2) %>%
  filter(gene.1 < gene.2) # To get unique combinations
```
  
```{r}
setwd(parent)

copy.box.plot(single.cell.progenitors.data)
#copy.box.plot(single.cell.progenitors.data, darkTheme)
concentration.box.plot(single.cell.progenitors.data)
#concentration.box.plot(single.cell.progenitors.data, darkTheme)

for (g in c(1,2,3)) {
  copy.pairwise.scatter.plot(
    single.cell.progenitors.data, 
    genes.pairwise$gene.1[g], genes.pairwise$gene.2[g],
    whiteTheme)
#  copy.pairwise.scatter.plot(
#    single.cell.progenitors.data, 
#    genes.pairwise$gene.1[g], genes.pairwise$gene.2[g],
#    darkTheme)
  concentration.pairwise.scatter.plot(
    single.cell.progenitors.data, 
    genes.pairwise$gene.1[g], genes.pairwise$gene.2[g],
    whiteTheme)
#  concentration.pairwise.scatter.plot(
#    single.cell.progenitors.data, 
#    genes.pairwise$gene.1[g], genes.pairwise$gene.2[g],
#    darkTheme)
}
```
  
source("26_cell_area.R")  
  
###########################################################################
###########################################################################
###########################################################################
###########################################################################
  
file$channel <- as.factor(file$channel)
levels(file$channel) <- genes

rect <- data.frame(xmin=threshold.area, xmax=Inf,
                   ymin=threshold.MaxInt, ymax=Inf,
                   channel=c(0,1,2))
rect$channel <- as.factor(rect$channel)
levels(rect$channel) <- genes

###########################################################################
###########################################################################

source("26_spot_area_maxInt.R")


source("26_TXnumber_vs.R")


source("26_barTXprob.R")


source("26_barTXprobExtrinsic.R")

###########################################################################
###########################################################################


nested <-
  total %>%
  group_by(day, image, cell.id, cell.area, channel, copy, density, TX.number) %>% nest() %>%
  group_by(day, image, cell.id, cell.area, channel) %>% nest() %>%
  group_by(day, image, cell.id, cell.area) %>% nest() %>%
  group_by(day) %>% nest()

myjson <- toJSON(nested, pretty = TRUE, auto_unbox = TRUE)

write(myjson, "single-RNAs_TX-RNAs_cell.json")

library(rjson)
json_data <- fromJSON(file="single-RNAs_TX-RNAs_cell.json")


###########################################################################
###########################################################################

source("26_boxTXmolecules.R")



source("26_TX2alleles.R")

###########################################################################
###########################################################################

plots <- c(hist.spots.area,
           hist.spots.MaxInt,
           p.spots.AreaVsMaxInt,
           p.TX.numberVsCopy,
           p.TX.numberVsDensity,
           p.TX.numberVsCellArea,
           bar.TX.prob,
           bar.TX.prob.extrinsic,
           box.TX.molecules,
           p.TX.2alleles.molecules.final)

###########################################################################
###########################################################################

#Heterogeneity analysis

var(spots.cell$copy.C1) / mean (spots.cell$copy.C1)

#Fano factor = variance / mean = sigma^2 / mean
FF.expressing <- data.frame(FF.C0=as.numeric(),
                            FF.C1=as.numeric(),
                            FF.C2=as.numeric())
i <- 1
for (d in days) {
  temp <- spots.cell.expressing[[1]]
  temp2 <- spots.cell.expressing[[2]]
  temp3 <- spots.cell.expressing[[3]]
  FF.expressing[i,] <- c(var(temp$copy.C0[temp$day==d]) / mean(temp$copy.C0[temp$day==d]),
                                 var(temp2$copy.C1[temp2$day==d]) / mean(temp2$copy.C1[temp2$day==d]),
                                 var(temp3$copy.C2[temp3$day==d]) / mean(temp3$copy.C2[temp3$day==d]))
  i <- i + 1
}



###########################################################################
###########################################################################

#Save extracted data to file

write.csv(spots.cell, "spots_cell.csv", row.names=FALSE)
write.csv(spots.cell.2, "spots_cell2.csv", row.names=FALSE)

write.csv(spots.cell.pro, "spots_cell_pro.csv", row.names=FALSE)
write.csv(spots.cell.2.pro, "spots_cell2_pro.csv", row.names=FALSE)

write.csv(TX.molecules, "TX_molecules.csv", row.names=FALSE)
